sdk/openai/Azure.AI.OpenAI/tests/OpenAITestBase.cs

```diff
@@ -16,138 +15,179 @@ namespace Azure.AI.OpenAI.Tests
+            public const string EmbeddingsModelName = "text-similarity-davinci-001";
+            public const string SubDomainPrefix = "sdk";
+            public static AzureLocation Location = AzureLocation.EastUS;
+        }
+
+        private static readonly object _deploymentIdLock = new object();
+
+        public string CompletionsDeploymentId { get => _completionsDeploymentId; }
+        public string EmbeddingsDeploymentId { get => _embeddingsDeploymentId; }
+        public OpenAIClientOptions TestClientOptions { get; set; } = new OpenAIClientOptions(OpenAIClientOptions.ServiceVersion.V2022_12_01);
+
+        private Uri _endpoint;
+        private AzureKeyCredential _apiKey;
+        private string _completionsDeploymentId;
+        private string _embeddingsDeploymentId;
 
         protected OpenAITestBase(bool isAsync, RecordedTestMode? mode = null) : base(isAsync, mode)
         {
             HeaderRegexSanitizers.Add(new Core.TestFramework.Models.HeaderRegexSanitizer("api-key", "***********"));
         }
 
-        private static readonly object _deploymentIdLock = new object();
+        protected OpenAIClient GetClient() => InstrumentClient(
+            new OpenAIClient(_endpoint, _apiKey, InstrumentClientOptions(TestClientOptions)));
+
+        protected OpenAIClient GetClientWithCredential() => InstrumentClient(
+            new OpenAIClient(_endpoint, TestEnvironment.Credential, InstrumentClientOptions(TestClientOptions)));
+
+        protected OpenAIClient GetClientWithCompletionsDeploymentId() => InstrumentClient(
+            new OpenAIClient(_endpoint, CompletionsDeploymentId, TestEnvironment.Credential, InstrumentClientOptions(TestClientOptions)));
+
         [SetUp]
         public void CreateDeployment()
         {
             if (Mode == RecordedTestMode.Playback)
             {
-                DeploymentId = Recording.GetVariable(DeploymentIdVariable, null);
-                EmbeddingsDeploymentId = Recording.GetVariable(EmbeddingsDeploymentIdVariable, null);
-                _endpoint = Recording.GetVariable(EndpointVariable, null);
+                _endpoint = new Uri(Recording.GetVariable(Constants.EndpointVariable, null));
+                _completionsDeploymentId = Recording.GetVariable(Constants.CompletionsDeploymentIdVariable, null);
+                _embeddingsDeploymentId = Recording.GetVariable(Constants.EmbeddingsDeploymentIdVariable, null);
+                _apiKey = new AzureKeyCredential("unused placeholder value for recordings");
             }
-
-            if ((Mode == RecordedTestMode.Live || Mode == RecordedTestMode.Record) && (DeploymentId == null || EmbeddingsDeploymentId == null))
+            else if (Mode == RecordedTestMode.Live || Mode == RecordedTestMode.Record)
             {
+                TestEnvironment.ThrowIfCannotDeploy();
+
                 lock (_deploymentIdLock)
                 {
-                    if (DeploymentId == null || EmbeddingsDeploymentId == null)
```

Comment:
I think we still want a 2nd if here to skip if another thread beat us inside the lock and took care of set up.  It could be that running the code twice is idempotent so this is just a performance enhancement, but its a good practice in general.

Reply:
Great point; I flipped this back around to move the check inside the lock.

Reply:
I think this should be good to go; per discussion, the `else -> lock -> if (use service)` chain should be just about equivalent to the top-level `else if` in this case, given we're already handling Playback earlier and don't expect anything other than `Live` and `Record` to proceed.